substrate/frame/nomination-pools/src/migration.rs

```diff
@@ -22,56 +22,246 @@ use sp_std::{collections::btree_map::BTreeMap, vec::Vec};
 
 #[cfg(feature = "try-runtime")]
 use sp_runtime::TryRuntimeError;
+pub mod v6 {
+	use super::*;
 
-pub mod v1 {
+	/// Wrapper over `MigrateToV6` with convenience version checks.
+	pub type VersionedMigrateV6<T> = frame_support::migrations::VersionedMigration<
+		5,
+		6,
+		MigrateToV6<T>,
+		crate::pallet::Pallet<T>,
+		<T as frame_system::Config>::DbWeight,
+	>;
+
+	/// This migration would restrict reward account of pools to go below ED by doing a named
+	/// freeze on all the existing pools.
+	pub struct MigrateToV6<T>(sp_std::marker::PhantomData<T>);
+
+	impl<T: Config> MigrateToV6<T> {
+		fn freeze_ed(pool_id: PoolId) -> Result<(), ()> {
+			let reward_acc = Pallet::<T>::create_reward_account(pool_id);
+			Pallet::<T>::freeze_min_balance(&reward_acc).map_err(|e| {
+				log!(error, "Failed to freeze ED for pool {} with error: {:?}", pool_id, e);
+				()
+			})
+		}
+	}
+	impl<T: Config> OnRuntimeUpgrade for MigrateToV6<T> {
```

Comment:
Would be a nice-to-have (but not a blocker) to have a unit test for this

Reply:
Not sure what exactly to test. The migration code uses a pallet fn `freeze_pool_deposit` which is tested as part of pallet unit tests, and its pretty simple migration other than that. There is a `post_upgrade` check though to test if there are any imbalances left in the pool.